hack/generator/pkg/astmodel/flags.go

```diff
@@ -0,0 +1,9 @@
+package astmodel
+
+type Flag string
```

Comment:
I feel like `Flag` is so generic - these have a targeted use case and wondering if we need a more targeted name as a result?

I guess they are `astmodel` `Flag`'s but still, there are a lot of possible types in `astmodel` which _could_ be flagged?

@Porges had mentioned `Metadata` as an alternative and I think I like something like `ObjectMetadata` better?

Reply:
They're supposed to be generic - and usable across a bunch of different scenarios.

Having `ObjectMetadata` implies fields like `IsAnyType`, `IsOneOfType` and `IsStorageType` which gets away from the idea of a generic mechanism.

Reply:
Yeah I'm cooling on metadata somewhat just because I don't know what "one metadata" is... whereas `Flag` is clearly the singular of the `Flags` collection.

Reply:
In terms of the level of genericism here, the part that feels awkward to me is that it's not clear at the flag definition point what places the author of the flag intended it to be valid for - but in practice they certainly intend it to be valid on some types but often not all. In order to understand flag usage you'd have to go find all of the locations it's checked in the code (which in practice may not be very difficult?). There's a disconnect between author intent and actual code rendered though which has to be made up by comments or something.

For example, putting `OneOfFlag` on a `TypeName` makes no sense (and wouldn't do anything).
Other flags (such at the storage one) make sense everywhere.

So it feels like flags have a `scope` or something, and we could limit flags to only be used in the scope they're allowed at.

```
type FlagScope int

// TODO: not sure if this is how you do a [flags] enum in Go...
const (
    Object = 0x1,
    Property = 0x2,
    All = Object | Property
    ...
)

type Flag struct {
    Scope FlagScope
    Name string 
}
```

Then individual types ensure they can't be flagged by flags that don't make sense

object_type AddFlag (for example)
```
func (o *ObjectType) AddFlag(flag Flag) *ObjectType {
    if !flag.Scope.HasFlag(Object) { // Here's the trouble with the generic flag name as well I think, because it collides with other contexts that may possibly show up in our code at a later date (such as a bitfield/flags enum).
        panic(fmt.Sprintf("Can't use a flag with %s scope on an Object", flag.Scope))
    }
    result := o.copy()
    result.flags[flag.Name] = flag
    return &result
}
```

This would allow an author to encode expected usage into the definition of the flag. It's still easy to broaden its usage in the future if so desired but it's explicit then.

Reply:
You raise some good points - but I fear this is veering into territory where it will just get over complicated. Maybe we need a different solution to "Flags".

Reply:
> I don't know what "one metadata" is...
A metadatum, obviously.